### ISSUE TITLE
Expose queue family index in `QueueFamilyProperties`

### ISSUE THREAD

**Issue by melvyn2** [vulkano-rs/vulkano#2039]
Creating a queue using a [`QueueCreateInfo`](https://docs.rs/vulkano/latest/vulkano/device/struct.QueueCreateInfo.html#) struct requires passing the queue family index. However, the [`queue_family_properties()`](https://docs.rs/vulkano/latest/vulkano/device/physical/struct.PhysicalDevice.html#method.queue_family_properties) method of a physical device returns an iterator of [`QueueFamilyProperties`](https://docs.rs/vulkano/latest/vulkano/device/struct.QueueFamilyProperties.html#), which do not store their index. 
The examples use the `enumerate()` method of iterators to get the index of the families in the `queue_family_properties` iterator, however this is unintuitive and requires casting the `usize` index to the `u32` that vulkano expects.
Adding the index field to the `QueueFamilyProperties` struct would make this much simpler and more intuitive, as would passing that struct directly to the `QueueCreateInfo` struct.

**Comment by Rua** [comment]
`queue_family_properties` returns a reference to a slice, not an iterator. And I think `enumerate` is the normal way of getting the index when iterating over a slice?

**Comment by melvyn2** [comment]
Oh I meant a slice instead of the first iterator, though my point mostly remains the same. The outdated guide uses this segment to create the queue:
```rust
use vulkano::device::{Device, DeviceCreateInfo, Features, QueueCreateInfo};

let (device, mut queues) = Device::new(
    physical,
    DeviceCreateInfo {
        // here we pass the desired queue families that we want to use
        queue_create_infos: vec![QueueCreateInfo::family(queue_family)],
        ..Default::default()
    },
)
.expect("failed to create device");
```
To me, this API (`QueueCreateInfo::family(queue_family)`) makes a lot more sense. If the user code is given a struct describing a Queue Family, and the API requires specifying a Queue Family, I think that it makes more sense to pass the struct the user is given, rather than making the user find the index of that struct. Towards this end, the `QueueCreateInfo` struct could either start accepting a vec of `QueueFamilyProperties` instead of indices (and add a constructor for `QueueFamilyProperties` using only a user-specified index as to allow manually specifying), or the struct could specify its own index that the user can pass to the `QueueCreateInfo`.

To put it another way, finding and specifying the index of a struct seems like something Vulkano should manage in the default case. Considering that one intent of this project is safety and input validation, using a raw index instead of a descriptor struct that already exists seems out of place.

**Comment by Rua** [comment]
But if all that's needed is the index, what good would providing the rest of the struct do?

**Comment by melvyn2** [comment]
For vulkano, nothing. For the user, it allows simply passing the struct that's given, as well as being more certain that what they're passing is a valid/existing queue family.

**Comment by trevex** [comment]
I have to agree that I believe the current way is less ergonomic than the snippet provided from the outdated guide by @melvyn2. 

Albeit this is a minor issue as this is boilerplate most users will only write once, I believe getting new users up and running quicker and reducing boilerplate in all provided examples is always desirable.

I have some additional thoughts on the matter:
* `queue_create_infos` maybe using a filter method directly in `QueueCreateInfo` would be helpful. Some filters could already be exposed, e.g. surface providing a method to filter for graphics and surface support.
* Each queue returned needs to be unique. Currently when working with multiple queues, I have to check that the index is not already planned to be used for another queue.
* If the queue index is "abstraced" away for the user, it might be worth considering using queues directly for command buiffer builders, so https://github.com/vulkano-rs/vulkano/blob/master/examples/src/bin/image/main.rs#L216 would directly take a queue and the queue argument of the subsequent execute could be omitted: https://github.com/vulkano-rs/vulkano/blob/master/examples/src/bin/image/main.rs#L291
